Polyfills/Window/Source/TimeoutDispatcher.cpp

```diff
@@ -0,0 +1,165 @@
+#include "TimeoutDispatcher.h"
+
+#include <cassert>
+
+namespace Babylon::Polyfills::Internal
+{
+    namespace
+    {
+        using TimePoint = std::chrono::time_point<std::chrono::steady_clock, std::chrono::microseconds>;
+
+        TimePoint Now()
+        {
+            return std::chrono::time_point_cast<std::chrono::microseconds, std::chrono::steady_clock>(std::chrono::steady_clock::now());
+        }
+    }
+
+    struct TimeoutDispatcher::Timeout
+    {
+        TimeoutId id;
+
+        // Make this non-shared when JsRuntime::Dispatch supports it.
+        std::shared_ptr<Napi::FunctionReference> function;
+
+        TimePoint time;
+
+        Timeout(TimeoutId id, Napi::Function function, TimePoint time)
+            : id{id}
+            , function{std::make_shared<Napi::FunctionReference>(Napi::Persistent(function))}
+            , time{time}
+        {
+        }
+
+        Timeout(const Timeout&) = delete;
+        Timeout(Timeout&&) = delete;
+    };
+
+    TimeoutDispatcher::TimeoutDispatcher(Babylon::JsRuntime& runtime)
+        : m_runtime{runtime}
+        , m_thread{std::thread{&TimeoutDispatcher::ThreadFunction, this}}
+    {
+    }
+
+    TimeoutDispatcher::~TimeoutDispatcher()
+    {
+        {
+            std::unique_lock<std::mutex> lk{m_mutex};
+            m_idMap.clear();
+            m_timeMap.clear();
+            m_shutdown = true;
+            m_condVariable.notify_one();
+        }
+
+        m_thread.join();
+    }
+
+    TimeoutDispatcher::TimeoutId TimeoutDispatcher::Dispatch(Napi::Function function, std::chrono::milliseconds delay)
+    {
+        if (delay.count() < 0)
+        {
+            delay = std::chrono::milliseconds{0};
+        }
+
+        const auto id = NextTimeoutId();
```

Comment:
This also uses the map to check if the id exists and thus is a race condition.

Reply:
It might be prudent to use two different mutexes for the condition variable and the maps.

Reply:
I'm going to need help with separate mutexes. I gave it a try [in this commit](https://github.com/docEdub/BabylonNative/commit/d6af3410da3866344996d493fe21cad37aa2c340) but it made most of the `setTimeout` unit tests time out.